Source/MayaLiveLinkPlugin.cpp

```diff
@@ -272,6 +272,188 @@ struct FStreamHierarchy
 	{}
 };
 
+/*
+*	Utility class for keeping track of user defined attributes from the
+*	channel box in Maya.
+*/
+class FStreamedUserDefinedAttributes
+{
+protected:
+	/*
+	*	This struct is used to cache each attribute MPlug for the streamed 
+	* 	user defined attributes. The struct contains member variables to
+	*	allow dirty-checking to help trigger a refresh of static data
+	*	whenever an attribute property is modified in the channel box 
+	*	in Maya.
+	*/
+	struct FAttributeInfo
+	{
+		// An MPlug is a connection which supports reading values and properties from object attributes
+		MPlug Plug;
+		inline float Value() { return Plug.asFloat(); }
+
+		// The cached partialName of the attribute
+		MString PartialName = "";
+
+		// Determines if an attribute should be streamed via Live Link (see IsPlugRelevantForStream)
+		bool bRelevantForStream = false;
+
+		// Determines if an attribute property has been modified by the user
+		bool HasDirtyProperties() const
+		{
+			return (PartialName != Plug.partialName()) || (bRelevantForStream != IsPlugRelevantForStream(Plug));
+		}
+	};
+
+	// The total number of attributes available on the root joint of the streamed hierarchy
+	int NumAttributesOnRootJoint = 0;
+
+	// The sub-set of attributes added by the user to the root joint
+	TArray<FAttributeInfo> UserDefinedAttributes;
+
+public:
+	/*
+	*	Determines if a user defined attribute should be streamed via Live Link.
+	*	Note that MPlug is used for querying properties and not MFnAttribute. 
+	*	Even though both has the same functions defined, only MPlug replicates the 
+	*	behavior of getAttr in MEL. MFnAttribute behaves as attributeQuery in MEL,
+	*	which do not report the state of the attribute that is needed.
+	*/
+	static bool IsPlugRelevantForStream(MPlug Plug)
+	{
+		if (Plug.isLocked())
+		{
+			return false;
+		}
+		else
+		{
+			return Plug.isKeyable() || Plug.isChannelBoxFlagSet();
+		}
+	}
+
+	/*
+	*	Returns the number of custom user defined attributes on the joint.
+	*	Attributes are index-based and custom attributes are added at the
+	*	end of the range. User defined attributes are marked isDynamic().
```

Comment:
That's from the Maya SDK documentation? Is it the same behavior across all Maya versions?

Reply:
I could not find any documentation in regards to user specific attributes. I had to print debug all states such as isDynamic() on each attribute until I could find one that was consistent for all user defined attributes. What I found was that all of them was appended at the end of the attribute index list. I have not confirmed if this works in later versions, but it was the only way I managed to separate them from the built in list.

Reply:
Might be better then to go through all attributes and fetch the dynamic ones just for the sake of potentially missing out some of them that would be in another order. If we react on attributeChanged messages would help keep track of that.